go/oasis-node/cmd/common/metrics/metrics.go

```diff
@@ -179,56 +313,533 @@ func (s *pushService) worker() {
 	}
 
 	svc := *service.NewBaseBackgroundService("metrics")
 
 	svc.Logger.Debug("Metrics Server Params",
-		"mode", metricsModePush,
+		"mode", MetricsModePush,
 		"addr", addr,
 		"job_name", jobName,
-		"instance_label", instanceLabel,
+		"labels", labels,
 		"push_interval", interval,
 	)
 
+	pusher := push.New(addr, jobName)
+	for k, v := range labels {
+		pusher = pusher.Grouping(k, v)
+	}
+
+	d, err := NewDiskService()
+	if err != nil {
+		return nil, err
+	}
+
+	m, err := NewMemService()
+	if err != nil {
+		return nil, err
+	}
+
+	c, err := NewCPUService()
+	if err != nil {
+		return nil, err
+	}
+
+	n, err := NewNetService()
+	if err != nil {
+		return nil, err
+	}
+
 	return &pushService{
 		BaseBackgroundService: svc,
-		pusher:                push.New(addr, jobName).Grouping("instance", instanceLabel),
+		pusher:                pusher,
 		interval:              interval,
+		dService:              d.(*diskService),
+		mService:              m.(*memService),
+		cService:              c.(*cpuService),
+		nService:              n.(*netService),
 	}, nil
 }
 
 // New constructs a new metrics service.
 func New(ctx context.Context) (service.BackgroundService, error) {
-	mode := viper.GetString(cfgMetricsMode)
+	mode := viper.GetString(CfgMetricsMode)
 	switch strings.ToLower(mode) {
-	case metricsModeNone:
+	case MetricsModeNone:
 		return newStubService()
-	case metricsModePull:
+	case MetricsModePull:
 		return newPullService(ctx)
-	case metricsModePush:
+	case MetricsModePush:
 		return newPushService()
 	default:
 		return nil, fmt.Errorf("metrics: unsupported mode: '%v'", mode)
 	}
 }
 
+type diskService struct {
+	service.BaseBackgroundService
+	sync.Mutex
+
+	dataDir string
+	// TODO: Should we monitor I/O of children PIDs as well?
+	pid      int
+	interval time.Duration
+
+	diskUsageGauge          prometheus.Gauge
+	diskIOReadBytesGauge    prometheus.Gauge
+	diskIOWrittenBytesGauge prometheus.Gauge
+}
+
+func (d *diskService) Start() error {
+	go d.worker()
+	return nil
+}
+
+// updateDiskUsage walks through dataDir and updates diskUsageGauge attribute.
+func (d *diskService) updateDiskUsage() error {
+	d.Lock()
+	defer d.Unlock()
+
+	// Compute disk usage of datadir.
+	var duBytes int64
+	err := filepath.Walk(d.dataDir, func(path string, info os.FileInfo, err error) error {
+		if err != nil {
+			return errors.Wrap(err, fmt.Sprintf("disk usage metric: failed to access file %s", path))
+		}
+		duBytes += info.Size()
+		return nil
+	})
+	if err != nil {
+		return errors.Wrap(err, fmt.Sprintf("disk usage metric: failed to walk directory %s", d.dataDir))
+	}
+	d.diskUsageGauge.Set(float64(duBytes))
+
+	return nil
+}
+
+// updateDiskUsage reads process info and updates diskIO{Read|Written}BytesGauge attributes.
+func (d *diskService) updateIO() error {
+	d.Lock()
```

Comment:
All calls to `updateIO` and `updateDiskUsage` come from the (probably single) worker loop and the calls are sequential so I think there is no need for the mutex?

Actually even if this was not true i don't see why a lock is needed?

Reply:
Actually this mutex is to prevent pusher (running in another thread) from *reading* diskService gauges while they are updated.

Reply:
But in that case the pusher should take a `{R}Lock()` somewhere before reading these gauges, maybe i just don't see where that happens.

Reply:
Also this leads me to believe that the prometheus client code is thread-safe: https://github.com/prometheus/client_golang/blob/ba7901740d8e73db921f90747908b7d92c3cc959/prometheus/gauge.go#L94-L96

Reply:
Yeah, you're right. Prometheus `Set` calls are thread-safe. I removed mutexes.